csrc/parallel_dimension_map.h

```diff
@@ -90,6 +92,8 @@ class ParallelDimensionMap {
 
   //! Set of parallel types that we are doing warp specialization on
   std::unordered_set<ParallelType> warp_specialized_types_;
+  std::unordered_set<ParallelType> ws_with_register_sharing_;
```

Comment:
Why do we need a `std::unordered_set` if we only allow `ws_with_register_sharing_` to contain only one value? i.e., Trigger an assertion in `ParallelDimensionMap::build` if there is multiple warp specialization parallel types.

Reply:
Why not replace `ws_with_register_sharing_` and `warp_specialization_padded_vals_` 
 with `std::optional<ParallelType, int64_t> ws_with_register_sharing_?

`std::nullopt` if there is not warp specialization.
`ParallelType` is warp_specialized.on parallel type.
`int64_t` is the padded value.

Reply:
> Why do we need a `std::unordered_set` if we only allow `ws_with_register_sharing_` to contain only one value? i.e., Trigger an assertion in `ParallelDimensionMap::build` if there is multiple warp specialization parallel types.

I probabally copied from `  std::unordered_set<ParallelType> warp_specialized_types_;`

Reply:
> Why not replace `ws_with_register_sharing_` and `warp_specialization_padded_vals_` with `std::optional<ParallelType, int64_t> ws_with_register_sharing_?
> 
> `std::nullopt` if there is not warp specialization. `ParallelType` is warp_specialized.on parallel type. `int64_t` is the padded value.

Revised to use 
```
  //! warp specialization with register sharing, keep track of
  //! the parallel type and the threads padding
  std::optional<ParallelType> ws_with_register_sharing_pt_;
  std::optional<int64_t> ws_with_register_sharing_pad_val_;
```